### PR TITLE
Honors Filter

### PR DESCRIPTION
**Pull request by SimChadha**
# Purpose

###### A brief description of the purpose of the PR's changes for the project.

Created Honors toggle that when activated, shows courses with at least one honors section

<br>

# Tickets

###### A link to any tickets this PR is associated with on Trello.

- https://trello.com/c/CQ602Xl2

# Contributors

###### Anyone who contributed to this PR for future reference.

- Simar Chadha
- Adam Koulopoulos

# Feature List

###### A list of more in-depth and technical changes, additions, deletions, and fixes this PR provides.

-
-
-

# Notes

###### A list of miscellaneous notes for this pull request, such as whether we need to add something later for production, or that it was time boxed, etc.

Currently uses a hard coded timer to allow the toggle to animate before activating the filter. This is not ideal and could be done better with async/await in the future. 

<br>

# Checklist

- [x] Filled out PR template :wink:
- [ ] Approved by designers
- [x] Is passing linting checks
- [x] Is passing tsc
- [x] Is passing existing tests
- [x] Has documentation and comments in code
- [ ] Has test coverage for code
- [x] Will have clear squash commit message

<br>
@sandboxnu/searchneu

### CODE REVIEW

File: components/ResultsPage/FilterPanel.tsx [hunk 1]
```diff
@@ -4,6 +4,8 @@ import CheckboxFilter from './CheckboxFilter';
 import DropdownFilter from './DropdownFilter';
 import { FilterOptions, FilterSelection, FILTERS_IN_ORDER } from './filters';
 import RangeFilter from './RangeFilter';
+import type { Option } from './filters';
```
Comment (andrewydai) [inline components/ResultsPage/FilterPanel.tsx]: remove to fix lint warrning

File: components/ResultsPage/ToggleFilter.tsx [hunk 2]
```diff
@@ -13,6 +12,18 @@ export default function ToggleFilter({
   selected,
   setActive,
 }: ToggleFilterProps): ReactElement {
+  let isActivated = selected;
+
+  function setActiveAfterAnimating() {
+    isActivated = !isActivated;
+    const expectedActivationState = isActivated;
+    setTimeout(() => {
+      if (expectedActivationState == isActivated) {
```
Comment (andrewydai) [inline components/ResultsPage/ToggleFilter.tsx]: should use triple equals just for style

File: components/ResultsPage/ToggleFilter.tsx [hunk 3]
```diff
@@ -21,16 +32,15 @@ export default function ToggleFilter({
...
-          id={id}
           type="checkbox"
+          onChange={setActiveAfterAnimating}
+          id={id}
+          defaultChecked={selected}
+          checked={selected}
         />
         <label className="react-switch-label" htmlFor={id}>
-          <span className="react-switch-button" />
+          <span className={`react-switch-button`} />
```
Comment (andrewydai) [inline components/ResultsPage/ToggleFilter.tsx]: this should probably be the original if there's no js expression in the string literal

File: components/ResultsPage/filters.ts [hunk 4]
```diff
@@ -73,20 +73,27 @@ const CLASSTYPE_SPEC: FilterSpec<'Checkboxes'> = {
...
-  order: 4,
+  order: 5,
 };
 const CLASSIDRANGE_SPEC: FilterSpec<'Range'> = {
   category: FilterCategories.Range,
   default: {},
   display: 'Course Number',
-  order: 5,
+  order: 6,
+};
```
Comment (andrewydai) [inline components/ResultsPage/filters.ts:89]: Kinda nit but can we move these in order of their filter order

File: components/ResultsPage/filters.ts [hunk 5]
```diff
@@ -73,20 +73,27 @@ const CLASSTYPE_SPEC: FilterSpec<'Checkboxes'> = {
...
+  display: 'Honors',
+  order: 4,
 };
 export const FILTER_SPECS = {
   nupath: NUPATH_SPEC,
   subject: SUBJECT_SPEC,
   campus: CAMPUS_SPEC,
   classType: CLASSTYPE_SPEC,
   classIdRange: CLASSIDRANGE_SPEC,
+  honors: HONORS_SPEC,
```
Comment (andrewydai) [inline components/ResultsPage/filters.ts]: same here with ordering

File: package.json [hunk 6]
```diff
@@ -39,6 +39,7 @@
     "jsonwebtoken": "^8.5.1",
     "jwt-decode": "^3.1.2",
     "lodash": "^4.17.20",
+    "migrate": "^1.7.0",
```
Comment (andrewydai) [inline package.json]: uh what is this I don't see it being used

File: pages/[campus]/[termId]/search/[query].tsx [hunk 7]
```diff
@@ -2,7 +2,7 @@
  * This file is part of Search NEU and licensed under AGPL3.
  * See the license file in the root folder for details.
  */
-import _ from 'lodash';
+import _, { filter } from 'lodash';
```
Comment (andrewydai) [inline pages/[campus]/[termId]/search/[query].tsx]: revert this, probably accidentally imported while typing filter lol

File: components/ResultsPage/ToggleFilter.tsx [hunk 8]
```diff
@@ -13,6 +12,18 @@ export default function ToggleFilter({
   selected,
   setActive,
 }: ToggleFilterProps): ReactElement {
+  let isActivated = selected;
+
+  function setActiveAfterAnimating() {
```
Comment (andrewydai) [inline components/ResultsPage/ToggleFilter.tsx]: I'm assuming this is to fix the issue of animation/async set stating that could cause weird states if the toggle is spammed. I think this is unnecessary after talking with some people. If there's a bug that this covers I could not replicate; pls lmk if there is so we can talk about it. In the mean time, this is what I had for an onChange and it seemed to work without bugs:
`const onChange = event => setActive(event.target.checked)`

**Review by andrewydai — CHANGES_REQUESTED** [review]
Left some comments

File: components/ResultsPage/ToggleFilter.tsx (hunk 8 above)
Comment (SimChadha) [reply components/ResultsPage/ToggleFilter.tsx]: I still seem to get a visual bug with that solution, but we can talk about it more

**Review by andrewydai — APPROVED** [review]
🚀

**Pull request merged**